downgrade

```diff
@@ -448,12 +436,33 @@ parse_options() {
 }
 
 cli() {
-  parse_options "$@"
+  local conf_args=()
+
+  # Get configuration and command-line arguments and parse everything
+  eval "conf_args=($(read_downgrade_conf))"
```

Comment:
I agree this is probably the only way to do it. I wonder if there are any security concerns with this here.

Reply:
Hmm yes I think there is a security risk here, for example if your `conf` file has malicious code such as:

```
--pacman-cache "$(sudo rm -rf /)"
```

This line would actually get executed during the `eval` call, which makes me slightly uncomfortable.

How about an alternative where we use regular expressions to split all non-commented lines in the configuration file into one option and one value (if present)? Will try and provide a sample in a bit,

Reply:
Happy to look at that code and see, but I wouldn't consider that possibility and immediate reason to change the implementation.

"The code in the conf will be `eval`-d" isn't itself a vulnerability. It requires that the code in the conf can somehow be attacker-controlled. And if an attacker can change `/etc/downgrade.conf` on you, I think you have bigger problems, right?

Reply:
Hmm I am reading a lot of advice to avoid `eval` online, but you do have a point as well.

I tried the regex approach but it seems like an overkill. I found this one simple approach using `xargs` (source: https://superuser.com/a/1519494):

```shell
$ a='"aString that may haveSpaces IN IT" bar foo "bamboo" "bam boo"'
$ printf "%s" "$a" | xargs -n 1 printf "%s\n"
aString that may haveSpaces IN IT
bar
foo
bamboo
bam boo
```

I tested this locally and it does not execute any commands during the reading process. WDYT?

Reply:
I think we can consider the `eval` save by passing the elements through `printf '%q'` first. Sorry, I had used this trick before and since forgotten. It actually makes things simpler in general, I think (no need for `tr` or messing with `$IFS`):

```console
safe_example() {
> cat <<'EOM'
> --foo bar
> --baz
> --bat "a hoy"
> EOM
> }
```

```console
evil_example() {
cat <<'EOM'
> --foo
> $(ls /etc)
> --bar
> EOM
> }
```

```console
[patrick@leo ops]$ eval "x=( $(safe_example | xargs printf '%q ') )"
[patrick@leo ops]$ printf '>%s<\n' "${x[@]}"
>--foo<
>bar<
>--baz<
>--bat<
>a hoy<
```

```console
[patrick@leo ops]$ eval "y=( $(evil_example | xargs printf '%q ') )"
[patrick@leo ops]$ printf '>%s<\n' "${y[@]}"
>--foo<
>$(ls<
>/etc)<
>--bar<
```

Reply:
Oh that's awesome! I will test this more and get back tomorrow. Wishing you a nice start of the week :)